enhancements/catalog-data-transfer-reduction.md

```diff
@@ -0,0 +1,450 @@
+transferring an entire catalog from disconnected to connected is equally
+as un-scalable as connected to disconnected. Instead, a simple *catalog snapshot*,
+a list of packages containing channel and bundle elements, is easy
+to both produce from in-cluster subscriptions and manipulate by hand:
+
+```yaml
+packages:
+- name: foo
+  channels:
+    - name: preview
+      head: foo.v0.1.0-beta.2
+      versions:
+      - 0.1.0-beta.0
+      - 0.1.0-beta.1
+      - 0.1.0-beta.2
+    - name: stable
+      head: foo.v0.1.0
+      versions:
+      - 0.1.0
+- name: bar
+- name: baz
+  channels:
+...
+```
+
+Head information is only required if performing a FromHeads diff.
+
+Optionally a DC describing the full disconnected catalog state can be provided instead of a snapshot,
+and can be transformed into the snapshot format internally.
+
+A standalone tool or new `opm` command will be written to build a catalog snapshot
+from a cluster by making [`ListBundles()`][listbundles] gRPC calls to all registries running.
+
+#### Merging catalogs
+
+Now that the diff `output` has been created, some mirroring command like `oc adm catalog mirror`
+will send the `output` (DC) and all referenced images to a disk mount.
+The mount is then transferred via meatspace to the disconnected side,
+and images are again mirrored from the mount to a registry.
+
+If there is no existing catalog in the cluster (fresh cluster),
+then a CatalogSource and set of Subscriptions are created and the process is complete.
+
+If one or more catalogs do exist in the cluster, then a new CatalogSource
+must be created for `output`. This is already possible to do.
+Since multiple catalogs can serve the same bundle, duplicated dependencies
+do not throw errors. All replaces will point to bundles in the cluster's catalog state
+by the algorithm.
+
+#### Automation
+
+Since this workflow is disjointed by the meatspace data transfer, it cannot be fully automated.
+However, the workflow stages on the connected and disconnected sides can be automated separately.
+
+On the connected side, the `opm diff` command could sit behind a server,
+receiving requests for diffs. On receiving a request, the server pulls `input_new`
+from its image registry and runs `opm diff` with the request specifying `input_old`,
+optionally `input_expand`, and the desired modes and options.
+The response is the generated diff, and any extra data required for further diff action
+(see [open question 1a](#open-questions-optional)).
+
+On the disconnected side:
+- `input_old` can be created periodically (cronjob) from the cluster’s state
+by the catalog snapshot tool and placed in some mounted directory, priming it for meatspace transfer.
+- `output` and some catalog data (potentially an entire CatalogSource)
+can be placed in some directory being watched by a daemon
+that will perform any necessary catalog add/merge behavior,
+then create/update a CatalogSource in-cluster.
+    - This may have security implications.
```

Comment:
It would be really nice to have this command create and push a merged catalog into a static location from which existing `CatalogSource` objects get their image.